src/lowering/desugar.ml

```diff
@@ -239,6 +242,53 @@ and lexp' = function
   | S.IdxE (e1, e2) -> I.IdxLE (exp e1, exp e2)
   | _ -> raise (Invalid_argument ("Unexpected expression as lvalue"))
 
+and rewrite_for_to_while p arr proj c0 c1 c2 e1 e2 =
+  (* for p in e1 e2 when e1 = ⟨...array-typed...⟩.vals()
+     ~~>
+     let arr = ⟨...array-typed...⟩ ;
+     let size = arr.size() ;
+     var indx = 0 ;
+     label l loop {
+       if indx < size
+       then { let p = arr[indx]; e2; indx += 1 }
+       else { break l }
+     } *)
+  let arr_typ = arr.note.note_typ in
+  let triv t = { note_typ = t; note_eff = T.Triv } in
+  let body arrb =
+    let size_exp =
+      exp { e1 with note = triv T.nat;
+                    it = CallE ({ c0 with note = triv T.(Func (Local, Returns, [], [], [nat]));
+                                          it = DotE (arrb, { proj with it = "size" }) },
+                                c1, c2) } in
+    let indx = fresh_var "indx" T.(Mut nat) in
+    let indexing_exp = exp
+           { note = triv T.(as_immut (as_array arr_typ));
+             at = arr.at;
+             it = IdxE (arrb, { arr with note = triv T.nat;
+                                         it = VarE { it = id_of_var indx;
+                                                     note = ();
+                                                     at = arr.at } }) } in
+    let size = fresh_var "size" T.nat in
+    blockE [ letD size size_exp
+           ; varD indx (natE Numerics.Int.zero)]
+      (whileE (primE (Ir_def.Ir.RelPrim (T.nat, LtOp))
+                 [varE indx; varE size])
+         (blockE [ letP p indexing_exp
+                 ; letP wildP e2]
+            (assignE indx
+               (primE (Ir_def.Ir.BinPrim (T.nat, AddOp))
+                  [ varE indx
+                  ; natE (Numerics.Int.of_int 1)])))) in
+  let arr_ir = exp arr in
+  match arr_ir.it with
+  | I.VarE _ -> body arr
+  | _ ->
+    let arrv = fresh_var "arr" arr_typ in
+    letE arrv arr_ir
+      (body { arr with it = VarE { it = id_of_var arrv; note = (); at = arr.at };
+                       note = triv arr.note.note_typ })
```

Comment:
This (let-bind if not trivial) is too much unrelated logic; I suggest to put that in `construct.ml` as
```
letNonTrivialE : exp -> (var -> exp) -> exp
```
or something like that.

And then maybe simply write
```
rewrite_for_to_while p arr proj c0 c1 c2 e1 e2 =
  letNonTrivialE (exp arr) (fn arr_ir ->
    …
  )
```
i.e. wrap it all in `letNonTrivialE`. This way code order matches evaluation order, is closer to the pseudocode in the comment, and the `arr_ir` is in scope, i.e. `body` doesn’t have to be a function.

Reply:
Yes, the name gets captured and the type mismatch detected by the IT lint:
```
Ill-typed intermediate code after Desugaring (use -v to see dumped IR):
test/run/optimise-for-array.mo:75.16-75.22: IR type error [M0000], subtype violation:
  Text
  [var Text]
```
I'll see if I can guard against this case.

Reply:
Don’t worry about explicitly guarding, just introduce a fresh name. We are doing that all over the place, e.g. in the await translation. If this becomes a problem we can introduce a separate pass that cleans up the code again, or be smarter in the backend. (But not before introducing a proper non-recursive `let` to the IR…)

Reply:
Heh, note to self: never argue with a red-team guy...

I agree that variable path compression should be eventually be done in the backend, but we don't have walkable def-use chains yet.

Consider me convinced.

Reply:
This should wrap it up: d8242d132. No more variable shortcuts.